File: 이건희/level2/호텔 대실.py
def solution(book_time):
    result = []
    book_time = sorted(book_time, key=lambda x: (x[0], x[1]))
    
    for i in book_time:
        time = []
        for j in i:
            hour, minute = map(int, j.split(':'))
            time.append(hour * 60 + minute)
            
        for j in range(len(result)):
            if result[j] <= time[0]:
                result[j] = time[1] + 10
                break
        else:
            result.append(time[1] + 10)
        result.sort()
        
    return len(result)

File: 이훈일/Programmers/Level2/호텔 대실.java
import java.util.*;

class Solution {
    public int solution(String[][] book_time) {
        int[][] times = new int[book_time.length][2];
        for(int i = 0; i < book_time.length; i++) {
            String[] start = book_time[i][0].split(":");
            String[] end = book_time[i][1].split(":");
            
            times[i][0] = Integer.parseInt(start[0]) * 60 + Integer.parseInt(start[1]);
            times[i][1] = Integer.parseInt(end[0]) * 60 + Integer.parseInt(end[1]) + 10;
        }
        
        Arrays.sort(times, new Comparator<int[]>() {
            @Override
            public int compare(int[] a1, int[] a2) {
                return a1[0] - a2[0];
            }
        });
        
        PriorityQueue<int[]> pq = new PriorityQueue<>(new Comparator<int[]> () {
            @Override
            public int compare(int[] a1, int[] a2) {
                return a1[1] - a2[1];
            }
        });
        for(int i = 0; i < times.length; i++) {
            int[] time = times[i];
            
            if(pq.isEmpty()) {
                pq.add(time);
            } else {
                int[] comp = pq.peek();
                    
                if(time[0] >= comp[1]) {
                    pq.poll();
                } 
                
                pq.add(time);
            }
        }
        
        return pq.size();
    }
}
